### PR TITLE
Fix k3s cloud-provider docs

### PR DESCRIPTION
**Pull request by futuretea**
Signed-off-by: futuretea <Hang.Yu@suse.com>

Depend on PR: https://github.com/harvester/cloud-provider-harvester/pull/24

**Related issues**
https://github.com/harvester/harvester/issues/2998

### CODE REVIEW

File: docs/rancher/cloud-provider.md [hunk 1]
```diff
@@ -46,69 +46,47 @@ When spinning up an RKE2 cluster using the Harvester node driver, select the `Ha
 
 ### Deploying to the K3s Cluster with Harvester Node Driver [Experimental]
 
-- Choose the Kubernetes version to be k3s and click the `Edit as YAML` button to config the K3s cluster YAML (For existing cluster, you can also click the `Edit YAML` button to update it):
+When spinning up a K3s cluster using the Harvester node driver, you can perform the following steps to deploy the harvester cloud provider:
 
-  ![](/img/v1.1/rancher/edit-k3s-cluster-yaml.png)
+1. Generate and inject cloud config for `harvester-cloud-provider`
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:51]:
```suggestion
1. Generate and inject cloud config for `harvester-cloud-provider`.
```

File: docs/rancher/cloud-provider.md [hunk 2]
```diff
@@ -46,69 +46,47 @@ When spinning up an RKE2 cluster using the Harvester node driver, select the `Ha
...
 
-- Edit K3s cluster YAML.
-    - Set `disable-cloud-provider: true` to disable default k3s cloud provider.
-    - Add `cloud-provider=external` to use harvester cloud provider.
+The cloud provider needs a kubeconfig file to work, a limited scoped one can be generated using the [generate_addon.sh](https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh) script available in the [harvester/cloud-provider-harvester](https://github.com/harvester/cloud-provider-harvester) repo.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:53]:
```suggestion
The cloud provider needs a kubeconfig file to work. A scope-limited one can be generated using the [generate_addon.sh](https://raw.githubusercontent.com/harvester/cloud-provider-harvester/master/deploy/generate_addon.sh) script available in the [harvester/cloud-provider-harvester](https://github.com/harvester/cloud-provider-harvester) repo.
```

File: docs/rancher/cloud-provider.md [hunk 3]
```diff
@@ -46,69 +46,47 @@ When spinning up an RKE2 cluster using the Harvester node driver, select the `Ha
...
-
-
-### Deploy external cloud provider
-Deploying external cloud provider is similar for both RKE2 and K3s based clusters.
-
-Once the in-tree cloud provider has been disabled by following the above steps, you can deploy the external cloud provider via:
+:::note
 
-  ![](/img/v1.1/rancher/external-cloud-provider-addon.png)
+The script depends on `kubectl` and `jq` to operate the Harvester cluster
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:57]:
```suggestion
The script depends on `kubectl` and `jq` to operate the Harvester cluster.
```

File: docs/rancher/cloud-provider.md (hunk 2 above)
Comment (vickyhella) [reply docs/rancher/cloud-provider.md:53]: Not sure if I understand `limited scoped` correctly. Please double check

File: docs/rancher/cloud-provider.md [hunk 4]
```diff
@@ -46,69 +46,47 @@ When spinning up an RKE2 cluster using the Harvester node driver, select the `Ha
...
-  namespace: kube-system
-spec:
-  targetNamespace: kube-system
-  bootstrap: true
-  repo: https://charts.harvesterhci.io/
-  chart: harvester-cloud-provider
-  version: 0.1.12
-  helmVersion: v3
-```
+The script needs access to the `Harvester Cluster` kubeconfig to work.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:59]:
```suggestion
The script needs access to the `Harvester Cluster` kubeconfig.
```

File: docs/rancher/cloud-provider.md [hunk 5]
```diff
@@ -46,69 +46,47 @@ When spinning up an RKE2 cluster using the Harvester node driver, select the `Ha
...
 
-The cloud provider needs a kubeconfig file to work, a limited scoped one can be generated using the `generate_addon.sh` script available in the [harvester/cloud-provider-harvester](https://github.com/harvester/cloud-provider-harvester) repo.
+The namespace needs to be the namespace in which the guest cluster will be created.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:61]:
```suggestion
The namespace needs to be the one in which the guest cluster will be created.
```

File: docs/rancher/cloud-provider.md [hunk 6]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+########## cloud-init user data ############
+write_files:
+- encoding: b64
+  content: <CONTENT>
+  owner: root:root
+  path: /etc/kubernetes/cloud-config
+  permissions: '0644'
+```
+
+Copy and paste the output below `cloud-init user data` to **Machine Pools >Show Advanced > User Data**.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:133]:
```suggestion
Copy the output below `cloud-init user data` and paste it to **Machine Pools >Show Advanced > User Data**.
```

File: docs/rancher/cloud-provider.md [hunk 7]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+
+![](/img/v1.1/rancher/cloud-config-userdata.png)
+
+3. Add the following `HelmChart` yaml of `harvester-cloud-provider` to **Cluster Configuration > Add-On Config > Additional Manifest**
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:137]:
```suggestion
3. Add the following `HelmChart` yaml of `harvester-cloud-provider` to **Cluster Configuration > Add-On Config > Additional Manifest**.
```

File: docs/rancher/cloud-provider.md [hunk 8]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+  bootstrap: true
+  repo: https://charts.harvesterhci.io/
+  chart: harvester-cloud-provider
+  version: 0.1.13
+  helmVersion: v3
+```
+
+![](/img/v1.1/rancher/external-cloud-provider-addon.png)
+
+4. Disable the in-tree cloud provider by
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:156]:
```suggestion
4. Disable the in-tree cloud provider as follows:
```

Comment (vickyhella) [reply docs/rancher/cloud-provider.md:156]: If we use `by` here, all the verbs below should be in the`ing` form

File: docs/rancher/cloud-provider.md [hunk 9]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+
+- Click the `Edit as YAML` button
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:158]:
```suggestion
- Click the `Edit as YAML` button.
```

File: docs/rancher/cloud-provider.md [hunk 10]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+
+![](/img/v1.1/rancher/edit-k3s-cluster-yaml.png)
+- Disable `servicelb` and Set `disable-cloud-controller: true` to disable default k3s cloud controller.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:161]:
```suggestion
- Disable `servicelb` and set `disable-cloud-controller: true` to disable the default k3s cloud controller.
```

File: docs/rancher/cloud-provider.md [hunk 11]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+```yaml
+    machineGlobalConfig:
+      disable:
+        - servicelb
+      disable-cloud-controller: true
+```
+
+- Add `cloud-provider=external` to use harvester cloud provider.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:169]:
```suggestion
- Add `cloud-provider=external` to use the Harvester cloud provider.
```

File: docs/rancher/cloud-provider.md [hunk 12]
```diff
@@ -140,13 +118,66 @@ preferences: {}
...
+          - cloud-provider=external
+          protect-kernel-defaults: false
 ```
 
-This cloud-config file can now be injected via the `user-data` available in the `advanced options` for the nodepool.
-  ![](/img/v1.1/rancher/cloud-config-userdata.png)
+![](/img/v1.1/rancher/k3s-cluster-yaml-content-for-harvester-cloud-provider.png)
 
-With these settings in place a K3s / RKE2 cluster should provision successfully while using the external cloud provider.
+With these settings in place a K3s cluster should provision successfully while using the external cloud provider.
```
Comment (vickyhella) [inline docs/rancher/cloud-provider.md:180]:
```suggestion
With these settings in place, a K3s cluster should be provisioned successfully using the external cloud provider.
```

**Review by vickyhella — COMMENTED** [review]
Just some nits, the rest LGTM!

**Review by vickyhella — APPROVED** [review]

File: docs/rancher/cloud-provider.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (futuretea) [reply docs/rancher/cloud-provider.md:137]: No, otherwise, the k3s cluster cannot be active and Apps cannot be used from the UI

**Review by guangbochen — APPROVED** [review]
LGTM, thanks. cc @WuJun2016, will also need to include some changes as the default option with k3s provider on the UI, thanks. e.g.,
- Set `disable-cloud-provider: true` to disable default k3s cloud provider.
- Add `cloud-provider=external` to use harvester cloud provider.

**Pull request merged**